core/sugartypes.ml

```diff
@@ -274,10 +274,11 @@ and bindingnode =
                    (Types.datatype * Types.quantifier option list) option)
                    * funlit) * Location.t * datatype' option * Position.t) list
   | Handler of Binder.with_pos * handlerlit * datatype' option
+  | Mutual of binding list
```

Comment:
An alternative design would be to use a richer structure than a list, say
```ocaml
type mutual_defs = {
  mut_funs : binding;
  mut_types : binding;
}

let add_mutual_binding ({ mut_types = Typenames ts; mut_funs = Funs fs } as block) binding =
  match WithPos.node binding with
  | Fun _ -> { block with mut_funs = Funs (binding :: fs) }
  | Type _ -> { block with mut_types = Typenames (binding :: ts) }
  | _ -> *raise a syntax error*
```
With the help of `add_mutual_binding` the parser could partition the bindings; if I am not mistaken that would render `flatten_bindings` above obsolete, and safe at least a post-traversal.

Note this design abuses the parser, but I think it is justified as the partitioning logic wouldn't be contained directly within the parser.

Reply:
I'm not entirely sure about this -- saving a traversal would be nice, as would removing the binding flattening, but did we not recently discuss at a meeting that the parser is already doing too much work?

Reply:
Yes we did, but my point is that while this design would shift the work to the parser, the parser is somewhat oblivious to that fact. It is sort of like using a smart constructor.

Reply:
I went with this design in the end. It does introduce complexity in the parser (there are some intricacies not appearing in the above sketch) -- but I came round to the opinion that the additional parser complexity was worth removing the `Mutual` node and therefore most of `desugarMutual` completely. Good suggestion -- thanks.